File: 0054-spiral-matrix/0054-spiral-matrix.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        vector<int> sporder;
        int starti = 0;
        int startj = 0;
        int n = matrix.size();
        int m = matrix[0].size();
        int i = 0 , j = 0 , k = 0;
        while(starti<n && startj < m){

            if (starti == n - 1) {
                for (int j = startj; j < m; j++)
                    sporder.push_back(matrix[starti][j]);
                break;
            }

            // Only one column left
            if (startj == m - 1) {
                for (int i = starti; i < n; i++)
                    sporder.push_back(matrix[i][startj]);
                break;
            }

            j = startj;
            i = starti;
            while(j<m-1){
                sporder.push_back(matrix[i][j]);
                k++;
                j++;
            }
            
            while(i<n-1){
                sporder.push_back(matrix[i][j]);
                k++;
                i++;
            }
            
            while(j>startj){
                sporder.push_back(matrix[i][j]);
                k++;
                j--;
            }

            while(i>starti){
                sporder.push_back(matrix[i][j]);
                k++;
                i--;
            }
            
            starti++;
            startj++;
            m--;
            n--;
        }
        return sporder;
    }
};

File: 0054-spiral-matrix/0054-spiral-matrix.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        result = []

        rows = len(matrix)
        cols = len(matrix[0])

        top = 0
        bottom = rows - 1
        left = 0
        right = cols - 1

        while top <= bottom and left <= right:

            # left -> right
            for j in range(left, right + 1):
                result.append(matrix[top][j])
            top += 1

            # top -> bottom
            for i in range(top, bottom + 1):
                result.append(matrix[i][right])
            right -= 1

            # right -> left
            if top <= bottom:
                for j in range(right, left - 1, -1):
                    result.append(matrix[bottom][j])
                bottom -= 1

            # bottom -> top
            if left <= right:
                for i in range(bottom, top - 1, -1):
                    result.append(matrix[i][left])
                left += 1

        return result